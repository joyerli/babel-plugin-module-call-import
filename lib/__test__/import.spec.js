const babel = require('@babel/core');
const _ = require('lodash');
const plugin = require('../index');

function testUnit(source,
  {
    components = {},
    callback = (libraryName, ...modules) => `${libraryName}/${_.kebabCase(modules[1])}`,
  } = {}) {
  const { code } = babel.transform(source, {
    filename: 'path/to/file.js',
    sourceType: 'module',
    plugins: [[plugin, {
      components: {
        common: callback,
        ...components,
      },
    }]],
  });
  expect(code).toMatchSnapshot();
}

test('命名空间导入', () => {
  testUnit(`
    import * as Ad from 'antd';

    console.log(Ad.Row);
    console.log(Ad.DatePicker);
  `, {
    components: {
      antd(libraryName, ...modules) {
        return {
          source: `${libraryName}/lib/${_.kebabCase(modules[1])}`,
          extras: [
            `${libraryName}/lib/${_.kebabCase(modules[1])}/style`,
          ],
        };
      },
    },
  });
});
test('默认导入', () => {
  testUnit(`
    import El from 'element-ui';

    export default {
      name: 'TestComponent',
      components: {
        [El.ButtonGroup.name]: El.ButtonGroup,
        [El.Button.name]: El.Button,
      },
      data() {
        return {};
      },
    };
  `, {
    components: {
      'element-ui': (libraryName, ...modules) => ({
        source: `${libraryName}/lib/${_.kebabCase(modules[1])}`,
        extras: [`${libraryName}/lib/theme-chalk/${_.kebabCase(modules[1])}.css`],
      }),
    },
  });
});
test('具名导入', () => {
  testUnit(`
    import { Row, DatePicker } from 'antd';

    console.log(Row);
    console.log(DatePicker);
  `);
});
test('具名和默认导入', () => {
  testUnit(`
    import { Row, default as Ad } from 'antd';

    console.log(Row);
    console.log(Ad.DatePicker);
  `, {
    components: {
      antd(libraryName, ...modules) {
        return {
          source: `${libraryName}/lib/${_.kebabCase(modules[1])}`,
        };
      },
    },
  });
});
test('定制支持深层次使用', () => {
  testUnit(`
    import common from 'common';

    console.log(common.utils.ajax.params.parse);
  `, {
    callback(libraryName, ...modules) {
      return {
        source: [libraryName, ...modules.map((mod) => _.kebabCase(mod))].join('/'),
        namespaceLevel: 3,
      };
    },
  });
});
test('当不定制更深层时，当作属性使用', () => {
  testUnit(`
    import common from 'common';

    console.log(common.baseMixin.title);
  `);
});
