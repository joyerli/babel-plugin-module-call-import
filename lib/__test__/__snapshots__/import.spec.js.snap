// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`具名和默认导入 1`] = `
"import _default from \\"antd/lib/date-picker\\";
import { Row } from 'antd';
console.log(Row);
console.log(_default);"
`;

exports[`具名导入 1`] = `
"import { Row, DatePicker } from 'antd';
console.log(Row);
console.log(DatePicker);"
`;

exports[`命名空间导入 1`] = `
"import _default2 from \\"antd/lib/date-picker\\";
import \\"antd/lib/date-picker/style\\";
import _default from \\"antd/lib/row\\";
import \\"antd/lib/row/style\\";
console.log(_default);
console.log(_default2);"
`;

exports[`定制支持深层次使用 1`] = `
"import _default from \\"common/common/utils/ajax/params/parse\\";
console.log(_default);"
`;

exports[`当不定制更深层时，当作属性使用 1`] = `
"import _default from \\"common/base-mixin\\";
console.log(_default.title);"
`;

exports[`默认导入 1`] = `
"import _default2 from \\"element-ui/lib/button\\";
import \\"element-ui/lib/theme-chalk/button.css\\";
import _default from \\"element-ui/lib/button-group\\";
import \\"element-ui/lib/theme-chalk/button-group.css\\";
export default {
  name: 'TestComponent',
  components: {
    [_default.name]: _default,
    [_default2.name]: _default2
  },

  data() {
    return {};
  }

};"
`;
